File: widgets/button.lua
local _context = getContext()
local _painter = _context:getPainter()
local _style = _painter:getStyle()

exports('button', function(text)
	_painter:setText(text)
	_painter:setTextOpts()

	local w = _context:getWidgetWidth() or _painter:calculateTextWidth() + _style.button.spacing * 2
	local h = _style.widget.height

	_context:beginDraw(w, h)

	_painter:setColor(_context:isWidgetHovered() and _style.color.hover or _style.color.widget)
	_painter:drawRect(w, h)

	_painter:setColor(_style.color.primary)
	_painter:move(_style.button.spacing, 0)
	_painter:drawText()

	_context:endDraw()

	return _context:isWidgetClicked()
end)

File: src/widgets/button.ts
import { getCurrentContext } from '../../index';
import { Color, Image } from '../common/types';
import { Style } from '../core/style';

export function declareExport() {
	globalThis.exports('button', function (text?: string): boolean {
		const context = getCurrentContext();
		const painter = context.getPainter();
		const style = painter.getStyle();

		painter.setText(text);

		const buttonProperties = style.getProperties('button');
		const font = buttonProperties.get<number>('font-family');
		const scale = buttonProperties.get<number>('font-size');

		painter.setTextOptions(font, scale);

		const w = context.getWidgetWidth() ?? painter.calculateTextWidth() + style.button.spacing * 2;
		const h = style.widget.height;

		context.beginDraw(w, h);

		const properties = context.isWidgetHovered() ? style.getProperties('button:hover') : buttonProperties;

		const backgroundImage = properties.tryGet<Image>('background-image');
		if (backgroundImage !== undefined) {
			const backgroundColor = properties.tryGet<Color>('background-color');
			painter.setColor(backgroundColor ?? Style.SPRITE_COLOR);
			painter.drawSprite(backgroundImage[0], backgroundImage[1], w, h);
		} else {
			painter.setColor(properties.get<Color>('background-color'));
			painter.drawRect(w, h);
		}

		painter.setColor(properties.get<Color>('color'));
		painter.move(
			style.button.spacing,
			(h - painter.calculateTextLineHeight(font, scale)) / 2 + style.widget.textOffset
		);
		painter.drawText();

		context.endDraw();

		return context.isWidgetClicked();
	});
}
